src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,259 @@
+			return $this->fully_supported_input;
+		}
+
+		$stack = array();
+
+		$p = new WP_HTML_Tag_Processor( $this->html );
+		while ( $p->next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+			$tag_name = $p->get_tag();
+
+			if ( ! $p->is_tag_closer() ) {
+				$element = WP_HTML_Spec::element_info( $tag_name );
+
+				$self_closes = $element::is_void || ( ! $element::is_html && $p->has_self_closing_flag() );
+				if ( ! $self_closes ) {
+					$stack[] = $tag_name;
+				}
+			} else {
+				if ( end( $stack ) === $tag_name ) {
+					array_pop( $stack );
+					continue;
+				}
+
+				$this->fully_supported_input = false;
+				return false;
+			}
+		}
+
+		$this->fully_supported_input = 0 === count( $stack );
+
+		return $this->fully_supported_input;
+	}
+
+	public function next_tag( $query = null ) {
+		if ( false === $this->fully_supported_input || false === $this->ensure_support() ) {
+			return false;
+		}
+
+		if ( 0 < count( $this->open_elements ) ) {
+			$element = WP_HTML_Spec::element_info( end( $this->open_elements ) );
+			// @TODO: Handle self-closing HTML foreign elements: must convey self-closing flag on stack.
+			if ( $element::is_void ) {
+				array_pop( $this->open_elements );
+			}
+		}
+
+		if ( false === parent::next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+			return false;
+		}
+
+		$tag_name = $this->get_tag();
+		$element = WP_HTML_Spec::element_info( $tag_name );
+
+		$self_closes = $element::is_void || ( ! $element::is_html && $this->has_self_closing_flag() );
+		if ( $self_closes ) {
+			$this->open_elements[] = $tag_name;
+			return true;
+		}
+
+		if ( $this->is_tag_closer() ) {
+			array_pop( $this->open_elements );
+		} else {
+			$this->open_elements[] = $tag_name;
+		}
+
+		return true;
+	}
+
+	public function next_sibling() {
+		if ( false === $this->fully_supported_input || false === $this->ensure_support() ) {
+			return false;
+		}
+
+		$starting_depth = count( $this->open_elements );
+
+		while ( $this->next_tag() ) {
+			$current_depth = count( $this->open_elements );
+
+			if ( ! $this->is_tag_closer() && $current_depth === $starting_depth ) {
+				return true;
+			}
+
+			if ( ! $this->is_tag_closer() && $current_depth < $starting_depth ) {
+				return false;
+			}
+		}
+
+		return false;
+	}
+
+	public function next_child() {
+		if ( false === $this->fully_supported_input || false === $this->ensure_support() ) {
+			return false;
+		}
+
+		$starting_depth = count( $this->open_elements );
+
+		while ( $this->next_tag() ) {
+			$current_depth = count( $this->open_elements );
+
+			if ( ! $this->is_tag_closer() && $current_depth === $starting_depth + 1 ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	private function find_closing_tag() {
+		$starting_depth = count( $this->open_elements );
+
+		while ( $this->next_tag() ) {
+			$current_depth = count( $this->open_elements );
+
+			if ( $this->is_tag_closer() && $current_depth < $starting_depth ) {
+				return true;
```

Comment:
Ditto here for the depth check I mentioned above

Reply:
this is all garbage that will go away